App/Services/UpdateController.cs

```diff
@@ -0,0 +1,308 @@
+        //
+        // Swift's Sparkle does not support verifying app cast signatures yet,
+        // but we use this functionality on Windows for added security against
+        // malicious release notes.
+        var checker = new Ed25519Checker(SecurityMode.Strict,
+            publicKey: _config.PublicKeyBase64,
+            readFileBeingVerifiedInChunks: true);
+
+        // Tell NetSparkle to store its configuration in the same directory as
+        // our other config files.
+        var appConfigDir = SettingsManagerUtils.AppSettingsDirectory();
+        var sparkleConfigPath = Path.Combine(appConfigDir, "updater.json");
+        var sparkleAssemblyAccessor = new AssemblyDiagnosticsAccessor(null); // null => use current executable path
+        var sparkleConfig = new JSONConfiguration(sparkleAssemblyAccessor, sparkleConfigPath);
+
+        _sparkle = new SparkleUpdater(_config.AppCastUrl, checker)
+        {
+            Configuration = sparkleConfig,
+            // GitHub releases endpoint returns a random UUID as the filename,
+            // so we tell NetSparkle to ignore it and use the last segment of
+            // the URL instead.
+            CheckServerFileName = false,
+            LogWriter = new CoderSparkleLogger(logger),
+            AppCastHelper = new CoderSparkleAppCastHelper(_config.ForcedChannel),
+            UIFactory = uiFactory,
+            UseNotificationToast = uiFactory.CanShowToastMessages(),
+            RelaunchAfterUpdate = true,
+        };
+
+        _sparkle.CloseApplicationAsync += SparkleOnCloseApplicationAsync;
+
+        // TODO: user preference for automatic checking. Remember to
+        //       StopLoop/StartLoop if it changes.
+#if !DEBUG
+        _ = _sparkle.StartLoop(true, UpdateCheckInterval);
+#endif
+    }
+
+    private static async Task SparkleOnCloseApplicationAsync()
+    {
+        await ((App)Application.Current).ExitApplication();
+    }
+
+    public async Task CheckForUpdatesNow()
+    {
+        if (_sparkle == null)
+        {
+            _ = new MessageWindow(
+                "Updates disabled",
+                "The built-in updater is disabled by policy.",
+                "Coder Desktop Updater");
+            return;
+        }
+
+        // NetSparkle will not open the UpdateAvailable window if it can send a
+        // toast, even if the user requested the update. We work around this by
+        // temporarily disabling toasts during this operation.
+        var coderFactory = _uiFactory as CoderSparkleUIFactory;
+        try
+        {
+            if (coderFactory is not null)
+                coderFactory.ForceDisableToastMessages = true;
+
+            await _sparkle.CheckForUpdatesAtUserRequest(true);
+        }
+        finally
+        {
+            if (coderFactory is not null)
+                coderFactory.ForceDisableToastMessages = false;
+        }
+    }
+
+    public ValueTask DisposeAsync()
+    {
+        _sparkle?.Dispose();
```

Comment:
Needs to also de-register itself as a handler from the UserNotifier, otherwise it can't be GC'd.

Alternatively, do we really need this to be disposable?